Eagerly release GPU resources when we lose the device.

bradwerth opened the pull request:
**Connections**
N/A

**Description**
This destroys buffer objects whenever we "lose the device", including at the end of the `device.destroy` process. Destruction of the buffers triggers reclamation of GPU memory, rather than waiting for the garbage collector to drop the buffer. It would be ideal to also destroy texture objects, but `texture.destroy` is not implemented, as noted in a TODO.

**Testing**
This is exercised by any bulk test execution, like the CTS, which relies on `device.destroy` (not waiting for the garbage collector) to free up GPU resources. Bulk test execution will have less failures due to full memory conditions.

<!-- 
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [X] Run `cargo fmt`.
- [X] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [X] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

jimblandy committed:
[naga wgsl-in] Automatic conversions for global `var` initializers.

jimblandy committed:
[naga] Rename abstract-types snapshots abstract-types-const.

This prepares for introducing a similar test for global variables.

jimblandy committed:
[naga wgsl-in] Automatic conversions for local `var` initializers.

Commit:
Eagerly release GPU resources when we lose the device.

xiaopengli89 committed:
check multi-planar format

Commit:
Updated changelog.

teoxoy committed:
[vk] remove (old) unused blocklist for dual source blending (#4758)

valaphee committed:
Remove DX11 backend (#4828)

cwfitzgerald commented on wgpu-core/src/device/resource.rs:
```diff
@@ -3317,7 +3321,8 @@ impl<A: HalApi> Device<A> {
 
         // 1) Resolve the GPUDevice device.lost promise.
-        let closure = self.lock_life().device_lost_closure.take();
+        let mut life_lock = self.lock_life();
```
We shouldn't be holding the life lock across external closure invocations - this could potentially deadlock.

valaphee committed:
[gl] add support for line and point polygon modes (#4836)

Co-authored-by: Nicolas Silva <nical@fastmail.com>

dependabot[bot] committed:
Bump futures-lite from 2.0.1 to 2.1.0 (#4820)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

daxpedda committed:
Expose shader validation (#4811)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

daniel-keitel committed:
[wgpu-hal] Inline RayQuery Support (#3507)

Co-authored-by: JMS55 <47158642+JMS55@users.noreply.github.com>
Co-authored-by: Ashley Ruglys <ashley.ruglys@gmail.com>
Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

DouglasDwyer committed:
Fix Javascript exception on repeated `BufferSlice::get_mapped_range` calls (#4726)

* Add reusable buffer mappings for WASM

* Run cargo fmt

* Update CHANGELOG.md

* Update web.rs

* Add documentation for WebBuffer struct

cwfitzgerald committed:
Remove `expose-ids` Feature (#4841)

* Remove expose-ids feature

* Changelog

cwfitzgerald committed:
Some minor cleanups (#4843)

nical committed:
Work around cbindgen issue

cbindgen spins out of control if wgpu_hal reexports a wgpu_types type using pub type in stead of pub use while keeping the same name.

teoxoy committed:
[spv-out] Decorate all non-uniform binding array access chains as `NonUniform`, not just buffer binding arrays.

Apply the `NonUniform` decoration to the results of all access chains rooted in binding arrays that use non-uniform values as indices, regardless of the binding array's element type and address space. Previously, Naga only decorated non-uniform access chains for binding arrays of buffers.

teoxoy committed:
Update multi-planar texture API (#4837)

TheoDulka committed:
Use `VK_EXT_robustness2` only when not using an outdated intel iGPU driver (#4602)

dependabot[bot] committed:
Bump once_cell from 1.18.0 to 1.19.0 (#4847)

Bumps [once_cell](https://github.com/matklad/once_cell) from 1.18.0 to 1.19.0.
- [Changelog](https://github.com/matklad/once_cell/blob/master/CHANGELOG.md)
- [Commits](https://github.com/matklad/once_cell/compare/v1.18.0...v1.19.0)

---
updated-dependencies:
- dependency-name: once_cell
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Commit:
Ensure the lock is not held when invoking the device_lost_closure.

bradwerth commented:
Ugh, I botched the rebase/merge and I'm not sure how to recover. I'll resubmit as a new PR.

bradwerth commented:
New, non-borked PR is https://github.com/gfx-rs/wgpu/pull/4851.
